rclpy/rclpy/node.py

```diff
@@ -1118,9 +1138,21 @@ def create_publisher(
 
         callback_group = callback_group or self.default_callback_group
 
+        failed = False
+        try:
+            final_topic = self.resolve_topic_or_service_name(topic)
+        except RuntimeError:
+            failed = True
+        if failed:
+            self._validate_topic_or_service_name(topic)
```

Comment:
There are some tests expecting an [rclpy.InvalidTopicNameException](https://github.com/ros2/rclpy/blob/8948499b4898b8ed13c21b77532c17761310e523/rclpy/rclpy/exceptions.py#L56) when an invalid topic name is provided.

> Also, the failed variable seems redundant. We could easily rewrite this as:

Yeah that's true. The boolean avoids a longer error message including both raised exceptions though.

Reply:
> The boolean avoids a longer error message including both raised exceptions though.

Makes sense.

> There are some tests expecting an rclpy.InvalidTopicNameException when an invalid topic name is provided.

It's tricky, because we're assuming that the reason `resolve_topic_or_service_name` fails is due to an invalid topic name. I guess this should be the case, barring any unexpected bugs introduced to `resolve_topic_or_service_name`.

Reply:
> It's tricky, because we're assuming that the reason resolve_topic_or_service_name fails is due to an invalid topic name.

That's a good point.
https://github.com/ros2/rclpy/pull/635/commits/751a6010aafd3613dac856bee18c7263414c9c19 should handle both cases okey.